mesonbuild/msetup.py

```diff
@@ -243,6 +243,7 @@ def _generate(self, env: environment.Environment) -> None:
                 intr.backend.generate()
             b.devenv.append(intr.backend.get_devenv())
             build.save(b, dumpfile)
+            universal.pickle_save(b.environment.binaries, os.path.join(env.get_scratch_dir(), 'binaries.dat'))
```

Comment:
b.environment.binaries is a meson object. Unpickling this reveals that a bunch of modules are imported in order to restore this state.

I was thinking, if one wanted to avoid loading everything the usual coredata loads, that you'd probably want to save this:

```
binaries_dict = {'host': b.environment.binaries.host.binaries, 'build': b.environment.binaries.build.binaries}
```

Reply:
Although hmm, to be fair there are 70 modules loaded by binaries.dat, and 200 loaded by coredata.dat

Reply:
Oh I see... I wasn't aware of that. This is the first time I mess around with the pickle module. Would you mind telling me how you get to count those imports?

If I understand correctly, your suggestion is to basically convert the `b.environment.binaries` object into a plain dictionary, right? So there will be no need for any imports when unpickling it.

Reply:
```python
>>> import pickle, sys
>>> old = set(sys.modules.keys())
>>> with open('meson-private/binaries.dat', 'rb') as f:
...     data = pickle.load(f)
... 
>>> new = set(sys.modules.keys())
>>> new - old
```

Yes, what I suggested would effectively convert it into a plain dictionary.

Reply:
Oh, ok, thanks.

So, I thought about adding a `to_dict` method (or `__iter__` or something) to `PerMachine` so that the keys of the resulting dictionary are standardized and controlled by only one place. But `PerMachine` is a generic class, so that would only work for the things in `build` and `host` that also somehow support that operation. Is that ok? Or I should implement a specific solution for this case?

Reply:
@eli-schwartz refactored it to pickle plain `dict`s instead of the `PerMachine` object. I went for an `as_plain_dict` method instead of `__iter__` because I would personally like to use `MachineChoice`s as keys, but then that would not be a plain dictionary and would require `import`s.

Reply:
That seems fine.